### ISSUE TITLE
Non-ASCII characters don't work at all as bind key or in chat

### ISSUE THREAD

**Issue by Tomtom-mygit** [Unvanquished/Unvanquished#595]
(This issue is a copy of the following forum topic: http://forums.unvanquished.net/viewtopic.php?f=6&t=1743.)

>“[...] non-ASCII characters don't work at all as bind key or in chat (but works in console, and if you activate debug, there is keypress debug message only for non-ASCII character while you are typing) [...]” — [illwieckz](http://forums.unvanquished.net/viewtopic.php?f=6&t=1743#p15432)

For exemple, I'm using an AZERTY keyboard on Kubuntu 14.10 with french translation and locale and most of my binds are working except ones related to the *(1)*, *(2)* and *(3)* keys in the image below. When I press them nothing happen and when I press them while chatting I got the following error messages in `daemon.log`:
```
^3Warning: ^7Rocket_FromQuake: Could not find keynum 34
^3Warning: ^7Rocket_FromQuake: Could not find keynum 40
^3Warning: ^7Rocket_FromQuake: Could not find keynum 95
```
and the letters don't print in the chat.

My binds in `autoexec.cfg` are:
```
[...]
teambind aliens & class level1
teambind aliens é class level1upg
teambind aliens " class level2
teambind aliens ' class level2upg
teambind aliens ( class level3
teambind aliens - class level3upg
teambind aliens è class level4
[...]
teambind humans & "sell weapons;buy +psaw;buy +rifle"
teambind humans é "sell weapons;buy +shotgun;buy +rifle"
teambind humans " "sell weapons;buy +lgun;buy +rifle"
teambind humans ' "sell weapons;buy +mdriver;buy +rifle"
teambind humans ( "sell weapons;buy +chaingun;buy +rifle"
teambind humans - "sell weapons;buy +flamer;buy +rifle"
teambind humans è "sell weapons;buy +prifle;buy +rifle"
teambind humans _ "sell weapons;buy +lcannon;buy +rifle"
[...]
```

<a href="http://i.imgur.com/Y4D4Bzh.jpg"><img src="http://i.imgur.com/Y4D4Bzh.jpg" title="source: imgur.com" /></a>

I tried escaping the characters, like [peace](http://forums.unvanquished.net/viewtopic.php?f=6&t=1743#p15427) told me:
```
teambind aliens \" class level2
```
but it didn't worked unfortunately.

Then [illwieckz](http://forums.unvanquished.net/viewtopic.php?f=6&t=1743#p15458) suggested
>“[...]since there is no keypress debug messages, I think it's not an autoexec.cfg issue, since Unvanquished seems to not know the key is pressed.

>Perhaps the bug comes with the SDL2 port but I don't know exactly when. In the past Unvanquished was able to bind some strange keys (with a better support than tremulous and other idtech3 derivatives), but now it's worse.”

**Comment by Viech** [comment]
All parts of our game should fully support Unicode, as it was a while ago. I think libRocket broke this. Poking @DolceTriade for input.

**Comment by illwieckz** [comment]
As you see: the ”é” character is not handled correctly:

![méh](http://dl.illwieckz.net/u/thomas-debesse/bazar/unvanquished/bugs/unvanquished_2015-02-15_184950_001.png)

**Comment by DolceTriade** [comment]
I can't imagine how librocket would break this. No, while our game is fully unicode aware for /text/ it has a hardcoded set of keys. If the key doesn't have a definition in keys.h, it won't be found. Another possibility is that you're binding it to the wrong key. I think there is  a cvar: in_keyboardDebug 1 which will allow you to see the internal representation of each key. Turn it on and try pressing the keys.

**Comment by illwieckz** [comment]
Just saying, it's almost impossible for french people to write in chat (using the T key), only the console is usable. While chatting outside the console, if you push a key for a non-ascii char, the chat is broken, some previous character disappear and you can't type more characters.

I will try `in_keyboardDebug 1` to see more.

**Comment by illwieckz** [comment]
From chat:

```
+ 0x33 "M" Q:0x6d(m) 
  0x33 "M" Q:0x6d(m) 
+ 0x1f "é" Q:0x00(0x00) 
  0x1f "é" Q:0x00(0x00) 
+ 0x0b "H" Q:0x68(h) 
  0x0b "H" Q:0x68(h) 
```

From console:

```
+ 0x33 "M" Q:0x6d(m) 
  0x33 "M" Q:0x6d(m) 
Debug: Execing command 'keyup 253590570 109 74561' 
+ 0x1f "é" Q:0x00(0x00) 
  0x1f "é" Q:0x00(0x00) 
+ 0x0b "H" Q:0x68(h) 
  0x0b "H" Q:0x68(h) 
Debug: Execing command 'keyup 253590570 104 76231' 
```

`0x00` can't be a valid bind !

**Comment by DolceTriade** [comment]
We don't have a key for it, so you can't bind it. Thanks for reporting it, we'll look into a more flexible system for keys.

**Comment by illwieckz** [comment]
The “é” key is just the “2” key, so it's one of the closest and most usefull key to bind…

Anyway, we can't chat in game, only on console, and as you see, it broke character encoding.

**Comment by Viech** [comment]
I can confirm weird behavior of the chat prompt even for the ASCII character `'`. It would often not be inserted and the chat prompt remains blocked until you hit backspace a number of times. Also see #610.

**Comment by Amanieu** [comment]
We could just use SDL2 keys directly instead of using a layer like keys.h

**Comment by Kangz** [comment]
I'm not too keen on using SDL2 keys directly as it would but SDL specific code in the rest of the engine (vs. only sdl_input)

**Comment by Viech** [comment]
What where we using pre-libRocket?

**Comment by DolceTriade** [comment]
The chatfield code is not unicode aware. I will fix this.

**Comment by DolceTriade** [comment]
This is a slightly bigger fix than I anticipated and likely going to be bug prone. I will have it in early next release cycle so the change can soak for a bit.

**Comment by Tomtom-mygit** [comment]
Unfortunately, the keys “é”, “"” and “è” (respectively *(1)*, *(2)* and *(3)* keys in the image joined with my first post) are still not bindable. For example, if as a Human holding a construction kit and having enough money I stand before an armoury and press:
- “&”, I will buy a painsaw;
- “é”, nothing will happen (I won’t buy a shotgun);
- “"”, nothing will happen (I won’t buy a lasgun);
- “’”, I will buy a mass driver;
- “(”, I will buy a chaingun;
- ....

Maybe @illwieckz can confirm this, since we use the same type of keyboard.

P.-S. – I don’t know if it’s related, but MOUSE4 and MOUSE5 button of my mouse are not bindable either.
```
teambind humans MOUSE4 itemact medkit
teambind humans MOUSE5 itemact grenade
```

Thank you for the game, keep up the good work guys!

**Comment by DolceTriade** [comment]
Do they work in chat though? I'm not sure we have an immediate solution for the keys part...

**Comment by slipher** [comment]
If you were really hardcore about maximizing your gameplay edge and needed to be able to bind all the keys on the keyboard, you could compile with [this patch](https://github.com/slipher/Unvanquished/commit/bcf810756ae9093c01414c6a4eed220448170bf1), but many of the keys would have the wrong name. To use a key, you would need to bind the command to the key in the same position on a QWERTY keyboard. For example use `2` for `è`, `a` for `q`, `q` for `a`, etc.

**Comment by DolceTriade** [comment]
Perhaps the ultimate solution is to remove our own K_ keys and just use SDL directly. Or we can add all the other unused keys to our own enum.

**Comment by Tomtom-mygit** [comment]
@DolceTriade: Yes they work in chat, but there are warning messages.

<a href="http://imgur.com/le9LU8J"><img src="http://i.imgur.com/le9LU8J.jpg?1" title="source: imgur.com" /></a>

@slipher: Thank you for this patch but I’ll wait a little bit in case a solution not involving the rewriting my binds would be found.

**Comment by Tomtom-mygit** [comment]
Unfortunately I fear that this bug report should be reopened. Indeed, when I type the first line of characters of my keyboard *&é"’(-è_çà)=* in the chat (alpha 41), I get the strange sequence of the following screenshot.

<a href="http://imgur.com/LEMvRLO"><img src="http://i.imgur.com/LEMvRLO.jpg" title="source: imgur.com" /></a>

Surprisingly, when I write the same sequence in the console, it is displayed correctly. However, writing the numbers of the numerical keypad in the console causes their display, but also for the 3 and the 9, respectively a scroll down and a scroll up.

<a href="http://imgur.com/BSGYOlC"><img src="http://i.imgur.com/BSGYOlC.jpg" title="source: imgur.com" /></a>

**Comment by DolceTriade** [comment]
:(
